### PR TITLE
Add compute_path_sinuosity to path metrics

### LINKED ISSUE #904: [Metrics] Add function for computing sinuosity (Benhamou 2004)
**Is your feature request related to a problem? Please describe.**
Sinuosity is a measure of the tortuosity of a random search path, combining the distribution of turning angles with step length statistics. It is one of the core trajectory complexity metrics identified in the [trajr paper](https://doi.org/10.1111/eth.12739) and was recommended for inclusion in movement#406 (see [Zulip discussion](https://neuroinformatics.zulipchat.com/#narrow/channel/406001-Movement/topic/Trajectory.20Complexity.20Metrics.20.28Issue.20.23406.2F.23835.29)).

**Describe the solution you'd like**
### Definition
 
The corrected sinuosity index (Benhamou 2004, Eq. 8) is defined as:
 
$$S = 2 \left[ \bar{p} \left( \frac{1 + \bar{c}}{1 - \bar{c}} + b^2 \right) \right]^{-1/2}$$
 
where:
- $\bar{p}$ = mean step length
- $\bar{c} = \frac{1}{n}\sum_{i=1}^{n} \cos(\theta_i)$ = mean cosine of turning angles
- $b = \mathrm{SD}(p_i) \, / \, \bar{p}$ = coefficient of variation of step length
 
**Range:** $[0, \infty)$, where higher values indicate more tortuous paths. A perfectly straight path has $S = 0$ (since $\bar{c} = 1$). A Brownian-like path has $S = 2 / \bar{p}$ (since $\bar{c} \to 0$).

### Alternative names
 
This metric is also known as: tortuosity index (in some disciplines), corrected sinuosity, Benhamou sinuosity. The original (uncorrected) version by Bovet & Benhamou (1988) uses a simpler formula $S = 2 / [p(1 - \bar{c})]$ but requires constant step length and assumes a wrapped-normal turning angle distribution. The corrected version is strictly more general and should be preferred.
 

### Pseudocode
 
```
function compute_sinuosity(data, nan_policy="ffill"):
    # 1. Compute turning angles (depends on movement#833)
    theta = compute_turning_angle(data)
 
    # 2. Compute step lengths
    displacements = compute_forward_displacement(data)
    step_lengths = compute_norm(displacements)
 
    # 3. Handle NaN values per nan_policy
    #    (follow compute_path_length convention)
 
    # 4. Compute summary statistics
    c_bar = nanmean(cos(theta))      # mean cosine of turning angles
    p_bar = nanmean(step_lengths)    # mean step length
    b = nanstd(step_lengths) / p_bar # CV of step length
 
    # 5. Compute sinuosity (Benhamou 2004, Eq. 8)
    S = 2 * (p_bar * ((1 + c_bar) / (1 - c_bar) + b**2)) ** (-0.5)
 
    return S
```
 
### Edge cases and `nan_policy`
 
- When $\bar{c} = 1$ (perfectly straight): the denominator $(1 - \bar{c})$ is zero, and $S = 0$.
- When all steps are NaN: return NaN.
- When step length is zero (stationary): turning angle is NaN (from movement#833), which propagates correctly.
- `nan_policy` should follow the convention established in `compute_path_length`. Niko noted on Zulip that "we may realise that the current policies don't cover what we need" — this metric may be a case where that becomes relevant.
 
### Proposed API
 
```python
def compute_sinuosity(
    data: xr.DataArray,
    nan_policy: Literal["ffill", "scale"] = "ffill",
    nan_warn_threshold: float = 0.2,
) -> xr.DataArray:
```
 
**Returns:** A scalar per individual/keypoint (like `compute_path_length`), with `time` and `space` dimensions removed.

### References
 
- **Primary:** Benhamou, S. (2004). How to reliably estimate the tortuosity of an animal's path: straightness, sinuosity, or fractal dimension? *J. Theoretical Biology* 229(2), 209-220. Eq. 8. https://doi.org/10.1016/j.jtbi.2004.03.016
- **Original formulation:** Bovet, P. & Benhamou, S. (1988). Spatial analysis of animals' movements using a correlated random walk model. *J. Theoretical Biology* 131(4), 419-433. https://doi.org/10.1016/S0022-5193(88)80038-9
- **trajr implementation:** `TrajSinuosity2` — McLean & Skowron Volponi (2018). https://doi.org/10.1111/eth.12739
 
### Context
 
Opened following discussion with Niko on [Zulip](https://neuroinformatics.zulipchat.com/#narrow/channel/406001-Movement/topic/Trajectory.20Complexity.20Metrics.20.28Issue.20.23406.2F.23835.29).

### PR DESCRIPTION
**Pull request by isha822**
Opening this as a draft to get early eyes on the core logic and math implementation (Benhamou 2004). 

Key implementation details:
* Added min_step_length to propagate down to compute_turning_angle for noise filtering.
* Reused internal _slice_and_validate and _warn_about_nan_proportion.
* Added NaN safety for stationary paths to prevent division-by-zero.

I will start adding the pytest suite for this once the core mathematical approach looks good.

# EDITED 2026-07-23 by @niksirbi 

Adds `compute_path_sinuosity` (Benhamou 2004's corrected sinuosity index) to `movement.kinematics.path`.

## Description

**What is this PR**

- [ ] Bug fix
- [x] Addition of a new feature
- [ ] Other

**Why is this PR needed?**

Sinuosity is a standard measure of path tortuosity. Unlike the straightness index $D/L$, it does not depend on the number of steps in the path, so it can be compared across trajectories of different durations. It complements the path metrics already available in `movement`. See #904.

**What does this PR do?**

- Adds `compute_path_sinuosity(data, nan_warn_threshold=0.2)`, implementing Eq. 8 of Benhamou (2004):

  $$S = 2\left[\bar{p}\left(\frac{1+\bar{c}}{1-\bar{c}} + b^{2}\right)\right]^{-1/2}$$

  where $\bar{p}$ is the mean step length, $\bar{c}$ the mean cosine of the turning angles, and $b$ the coefficient of variation of step length. A perfectly straight path gives $S = 0$; higher values mean more tortuous movement.
- Reuses the existing `_segment_lengths`, `compute_turning_angle` and `_warn_about_nan_proportion` helpers rather than recomputing anything.
- Generalises the private `_validate_time_points` helper with a `min_points` argument (default `2`, so existing callers are unaffected). Sinuosity passes `min_points=3`, since at least one turning angle is required.
- Deliberately does **not** expose `min_step_length` (per #957 and the 2026-06-12 community call) or a `nan_policy`. NaN positions propagate to NaN steps and turning angles, and the summary statistics are computed over the remaining valid samples. Time windows are selected with `data.sel(time=slice(start, stop))` rather than dedicated arguments.
- Exports the function from `movement.kinematics`.

## References

- Closes #904
- Related to #406
- Sibling metric PR #1032 (`compute_maximum_expected_displacement`), which shares the same turning-angle machinery.

Benhamou, S. (2004). How to reliably estimate the tortuosity of an animal's path: straightness, sinuosity, or fractal dimension? *Journal of Theoretical Biology*, 229(2), 209-220. https://doi.org/10.1016/j.jtbi.2004.03.016

## How has this PR been tested?

New tests in `tests/test_unit/test_kinematics/test_path.py`, all passing locally alongside the existing suite:

- **Known closed-form values**: straight path → `0.0`; stationary path → `NaN`; a unit-step 90° zigzag → `2.0`.
- **Units**: the same zigzag scaled by 4 → `1.0`, confirming that sinuosity scales as $1/\sqrt{\text{length}}$.
- **Missing data**: a zigzag with one missing position still returns `2.0`, verifying that statistics are computed over the remaining valid samples; a fully missing track warns and returns `NaN`.
- **Validation**: `ValueError` for fewer than 3 time points and for 3D input.
- **Output contract**: `time` and `space` are dropped while other dimensions are preserved; `name` and `long_name` are set.
- **Robustness**: realistic direction reversals with variable stride, and a large tracking outlier, both return finite positive values.
- Sinuosity was also added to the shared `test_path_metrics_across_time_ranges` parametrisation, so it is exercised across the same pre-sliced time ranges as the other path metrics. The shared error matcher was relaxed to be metric-agnostic, since the required minimum differs per metric.

Existing functionality is unchanged: `_validate_time_points` keeps its previous behaviour for all current callers via the default `min_points=2`.

## Is this a breaking change?

No. Only additions, plus a new keyword argument with a backwards-compatible default on a private helper.

## Does this PR require an update to the documentation?

No manual changes needed. The function has a full numpydoc docstring (maths, notes on NaN handling and units, references, examples), and the API reference page is generated automatically from the `movement.kinematics` exports.

## Checklist:

- [x] The code has been tested locally
- [x] Tests have been added to cover all new functionality
- [x] The documentation has been updated to reflect any changes
- [x] The code has been formatted with [pre-commit](https://pre-commit.com/)

### CODE REVIEW

File: movement/kinematics/path.py [hunk 1]
```diff
@@ -9,9 +9,14 @@
...
 from typing import Literal
 
+import numpy as np
 import xarray as xr
 
-from movement.kinematics.kinematics import compute_backward_displacement
+from movement.kinematics.kinematics import (
+    compute_backward_displacement,
+    compute_forward_displacement,
+    compute_turning_angle,
```
Comment (niksirbi) [inline movement/kinematics/path.py]: compute_turning_angle is in `movement.kinematics.orientation` not in `movement.kinematics.kinematics`. You can either import from the correct specific submodule, or else import all 3 functions from the higher-level module `movement.kinematics` (both should work I think.)

File: movement/kinematics/path.py [hunk 2]
```diff
@@ -195,6 +200,126 @@ def compute_path_straightness(
...
+    # Sinuosity specifically needs 3 points for angles
+    n_time = data.sizes["time"]
+    if n_time < 3:
+        raise logger.error(
+            ValueError(
+                "At least 3 time points are required to compute sinuosity "
+                f"(got {n_time}). A minimum of two consecutive displacements "
+                "is needed to produce one turning angle."
+            )
+        )
```
Comment (niksirbi) [inline movement/kinematics/path.py]: I wonder if this is a good chance to modify `_slice_and_validate` so it accepts `min_points` kwarg (defaulting to 2). Then we could fold all this validation into one helper.

File: movement/kinematics/path.py [hunk 3]
```diff
@@ -195,6 +200,126 @@ def compute_path_straightness(
...
 
+def compute_path_sinuosity(
+    data: xr.DataArray,
+    min_step_length: float = 0.0,
+    start: float | None = None,
+    stop: float | None = None,
+    nan_policy: Literal["ffill", "scale"] = "ffill",
+    nan_warn_threshold: float = 0.2,
+) -> xr.DataArray:
+    r"""Compute the sinuosity of a path (Benhamou 2004).
```
Comment (niksirbi) [inline movement/kinematics/path.py]: Re the docsting, I would add the following sections:

- See Also: (linking to underlying function `compute_turning_angle`, and to the existing related metric `compute_path_straightness`)
- Notes: containing the reference and the math equations, and also a short note about sinuosity having units of 1/sqrt(length), and therefore interpretation depends on position units.
- Examples: similar to the other functions in this file (tackle only after the function signature is settled)

File: movement/kinematics/path.py [hunk 4]
```diff
@@ -195,6 +200,126 @@ def compute_path_straightness(
...
+    elif nan_policy != "scale":
+        raise logger.error(
+            ValueError(
+                f"Invalid value for nan_policy: {nan_policy}. "
+                "Must be one of 'ffill' or 'scale'."
+            )
+        )
+
+    # --- Step lengths -------------------------------------------------------
+    step_lengths = compute_norm(compute_forward_displacement(data))
```
Comment (niksirbi) [inline movement/kinematics/path.py]: after this line I would add:

```python
step_lengths = step_lengths.where(step_lengths > min_step_length)
```

The current implementation applies `min_step_length` assymetrically. Small steps are masked in `turning_angles` (so exlcuded from $\bar{c}$) but not in `step_lengths` (kept it $\bar{p}$ and $\bar{b}$).

Masking them in both makes sense, because the intention of `min_step_length` in `compute_turning_angle` is to exlude steps we consider as 'noise'.

File: movement/kinematics/path.py [hunk 5]
```diff
@@ -195,6 +200,126 @@ def compute_path_straightness(
...
+
+    if nan_policy == "ffill":
+        data = data.ffill(dim="time")
+    elif nan_policy != "scale":
+        raise logger.error(
+            ValueError(
+                f"Invalid value for nan_policy: {nan_policy}. "
+                "Must be one of 'ffill' or 'scale'."
+            )
+        )
```
Comment (niksirbi) [inline movement/kinematics/path.py]: I don't think that these NaN policies (originally made for path length) make much sense here.
I would remove the `nan_policy` kwarg entitely, and just let NaNs propagate, but compute all the statistics with `skipna=True`. This effectively keeps the current behaviour of the 'scale' branch, but it's not worth calling it a 'policy' here.

I would keep `nan_warn_threshold` as a still useful sanity check. Optionally you can add a paragraph to 'Notes' like this:

> NaN positions propagate to NaN step lengths and turning angles; the statistics are computed over the remaining valid samples.

With the above in mind, the argument order in the function signature should be:

`data, start, stop, nan_warn_threshold, min_step_length`

File: movement/kinematics/path.py [hunk 6]
```diff
@@ -195,6 +200,126 @@ def compute_path_straightness(
...
+    angle_term = xr.where(
+        is_straight,
+        0.0,
+        (1.0 + c_bar) / (1.0 - c_bar),
+    )
+
+    result = 2.0 * (p_bar_safe * (angle_term + b**2)) ** -0.5
+
+    result = xr.where(is_straight, 0.0, result)
+    result = xr.where(is_stationary, np.nan, result)
```
Comment (niksirbi) [inline movement/kinematics/path.py]: I think this final `xr.where(is_stationary, np.nan, result)` is redundant. `p_bar_safe` is already NaN where stationary, so it should propagate to the `result` anyway.

**Review by niksirbi — CHANGES_REQUESTED** [review]
Thanks @isha822. 

I had a first look at this PR and left some comments that we should address first before proceeding.

**Comment by isha822** [comment]
@niksirbi 

I've updated the PR to address all your review comments.

I also pulled in the latest changes from main and resolved the merge conflicts, so this is now fully synced with the recent start/stop removals and the new compute_directional_change function. 

Could you take a quick look at the updated core logic? If the math and structure look good to you now, I'll go ahead and push the updated test suite!

File: movement/kinematics/path.py [hunk 7]
```diff
@@ -393,9 +396,173 @@ def compute_directional_change(
...
+
+    return result
+
+
 def _validate_time_points(
     data: xr.DataArray,
     metric_name: str,
+    min_points: int = 2,
 ) -> xr.DataArray:
     """Validate dims/coords and require at least 2 time points.
```
Comment (niksirbi) [inline movement/kinematics/path.py]:
```suggestion
    """Validate dims/coords and require at least ``min_points`` time points.
```

File: movement/kinematics/path.py [hunk 8]
```diff
@@ -405,6 +572,8 @@ def _validate_time_points(
         Position data with ``time`` and ``space`` dimensions.
     metric_name : str
         Used in the error message when there are fewer than 2 time points.
```
Comment (niksirbi) [inline movement/kinematics/path.py]:
```suggestion
        Used in the error message when there are fewer than ``min_points``
        time points.
```

File: movement/kinematics/path.py [hunk 9]
```diff
@@ -414,7 +583,7 @@ def _validate_time_points(
     """
     validate_dims_coords(data, {"time": [], "space": []})
     n_time = data.sizes["time"]
-    if n_time < 2:
+    if n_time < min_points:
         raise logger.error(
             ValueError(
                 "At least 2 time points are required to compute "
```
Comment (niksirbi) [inline movement/kinematics/path.py]:
```suggestion
                f"At least {min_points} time points are required to compute "
```

File: movement/kinematics/path.py (hunk 3 above)
Comment (niksirbi) [reply movement/kinematics/path.py]: No need to mention the paper in the first sentence. It's enough to reference it later:

```suggestion
    r"""Compute the sinuosity of a path.
```

File: movement/kinematics/path.py [hunk 10]
```diff
@@ -393,9 +396,173 @@ def compute_directional_change(
...
+    data: xr.DataArray,
+    nan_warn_threshold: float = 0.2,
+    min_step_length: float = 0.0,
+) -> xr.DataArray:
+    r"""Compute the sinuosity of a path (Benhamou 2004).
+
+    Sinuosity quantifies the tortuosity of a random search path by
+    combining turning angle statistics with step-length variability.
+    Higher values indicate more tortuous movement. A perfectly straight
+    path has S = 0.
```
Comment (niksirbi) [inline movement/kinematics/path.py]: Small tweaks:

```suggestion
    Sinuosity (S) quantifies the tortuosity of a path by combining
    turning angle statistics with step-length variability. Higher
    values indicate more tortuous movement. A perfectly straight
    path has S = 0.
``

File: movement/kinematics/path.py [hunk 11]
```diff
@@ -393,9 +396,173 @@ def compute_directional_change(
...
+
+    The corrected sinuosity index (Benhamou 2004, Eq. 8) is defined as:
```
Comment (niksirbi) [inline movement/kinematics/path.py]: You can refer to the paper you've already defined in 'References' using the `[1]_` syntax:
```suggestion
    The corrected sinuosity index (Eq. 8 in [1]_) is defined as:
```

File: movement/kinematics/path.py (hunk 4 above)
Comment (niksirbi) [reply movement/kinematics/path.py]: This is an opportunity to re-use the `_segment_lengths` helper we already have (was added after you opened the initial draft PR).

`_segment_lengths` computs step lengths based on the backwards displacement (as opposed to forward displacement), and I actually think that's better for readability and consistency. `compute_turning_angle` internally uses backward displacement, and the `min_step_length` masking inside it operates on backward displacement magnitudes. Using the same basis here in the sinuosity function makes it easier to reason about which physical steps are masked.

```suggestion
    step_lengths = _segment_lengths(data)
```

File: movement/kinematics/path.py [hunk 12]
```diff
@@ -393,9 +396,173 @@ def compute_directional_change(
...
+    # applied inside compute_turning_angle so noise steps are excluded
+    # from p_bar, c_bar, and b consistently.
+    step_lengths = step_lengths.where(step_lengths > min_step_length)
+
+    # --- Turning angles -----------------------------------------------------
+    theta = compute_turning_angle(data, min_step_length=min_step_length)
+
+    # --- Summary statistics (NaN-aware) -------------------------------------
+    p_bar = step_lengths.mean(dim="time", skipna=True)
+    c_bar = np.cos(theta).mean(dim="time", skipna=True)  # type: ignore
```
Comment (niksirbi) [inline movement/kinematics/path.py]: You can avoid the `type: ignore` by using `xr.apply_ufunc` which makes `mypy` aware that the returned type will be an xarray data array.
```suggestion
    c_bar = xr.apply_ufunc(np.cos, theta).mean(dim="time", skipna=True)
```

File: movement/kinematics/path.py [hunk 13]
```diff
@@ -393,9 +396,173 @@ def compute_directional_change(
...
+
+    Notes
+    -----
+    Step lengths are computed as the norm of forward displacement vectors
+    via :func:`~movement.utils.vector.compute_norm` and
+    :func:`~movement.kinematics.compute_forward_displacement`.
+    Turning angles are computed via :func:`compute_turning_angle`,
+    which internally composes
+    :func:`~movement.kinematics.compute_backward_displacement` and
+    :func:`~movement.utils.vector.compute_signed_angle_2d`.
```
Comment (niksirbi) [inline movement/kinematics/path.py]: Should be changed to backward displacement if you accept my suggestion above about `_segment_lengths()`.
No need to also cover internal compositions inside `compute_turning_angle`. If people are interested, they can go to that function's definition.

```suggestion
    Step lengths are computed as the norm of backward displacement vectors
    via :func:`~movement.utils.vector.compute_norm` and
    :func:`~movement.kinematics.compute_backward_displacement`.
    Turning angles are computed via :func:`compute_turning_angle`.
```

File: movement/kinematics/path.py [hunk 14]
```diff
@@ -393,9 +396,173 @@ def compute_directional_change(
...
+        0.0,
+        (1.0 + c_bar) / (1.0 - c_bar),
+    )
+    result = 2.0 * (p_bar_safe * (angle_term + b**2)) ** -0.5
+
+    result = xr.where(is_straight, 0.0, result)
+
+    result.name = "sinuosity"
+    result.attrs["long_name"] = "Path Sinuosity"
+    result.attrs["units"] = "1/sqrt(length)"
```
Comment (niksirbi) [inline movement/kinematics/path.py]: We don't yet use the `units` attribute for any other metric. I would leave it out for consistency.

File: movement/kinematics/path.py [hunk 15]
```diff
@@ -393,9 +396,173 @@ def compute_directional_change(
...
+    different spatial units.
+
+    **Edge cases**:
+
+    - :math:`\bar{c} = 1` (perfectly straight path): the denominator
+      :math:`(1-\bar{c})` is zero; S is explicitly set to ``0.0``
+      via :func:`xarray.where` before the division is evaluated.
+    - :math:`\bar{p} = 0` (entirely stationary track): S is NaN,
+      propagated from the safe division guard on :math:`\bar{p}`.
+    - All steps NaN: returns NaN.
```
Comment (niksirbi) [inline movement/kinematics/path.py]: I'd include somewhat less implementation detail for each edge case. It's more important to say what is returned in each case; people interested in the how can always check the code itself.
```suggestion
    - :math:`\bar{c} = 1` (perfectly straight path): the denominator
      :math:`(1-\bar{c})` is zero; S is explicitly set to ``0.0``
      before the division is evaluated.
    - :math:`\bar{p} = 0` (entirely stationary track): S is NaN.
    - All steps NaN: returns NaN.
```

**Review by niksirbi — COMMENTED** [review]
Thanks for keeping an eye on this PR and on updating it based on the latest change in main @isha822 !

I had a look at the current implementation and left some comments.

I think it's time to move onto writing tests for this function, in `test_path.py`. There could be subtle implementation issues we've both missed, but tests are the perfect means to reveal them. I would pay particular attention to checking that the edge cases are resolved as the docstring claims.

Feel free to take your time with the tests. More important to get this right than to get it done fast.

**Comment by isha822** [comment]
Thank you @niksirbi for the review. The suggestions are very helpful.

I'll work on incorporating them, cover the relevant edge cases, and add checks to catch potential anomalies. I'll update the PR as soon as I can.

**Comment by isha822** [comment]
Hey @niksirbi 

Just resolved the merge conflicts and synced all your suggestions from earlier!   

You'll notice the CI is failing right now. Aside from a few linting/formatting quirks I need to smooth out from the merge, the main test suite is failing on one specific biological edge case I added: test_path_sinuosity_perfect_reversals. 

I was thinking about real-world scenarios and realized there's a math quirk. If we track an animal perfectly pacing back and forth with a constant stride (A -> B -> A -> B), the mean turning angle cosine becomes exactly -1 and the step variation is 0. When that hits Eq 8, it throws a divide-by-zero warning and spits out inf.

What do you think? is this test valid?, let me know if im missing something or got anything wrong.

File: movement/kinematics/path.py [hunk 16]
```diff
@@ -510,18 +512,181 @@ def compute_path_deviation(
...
+    nan_warn_threshold: float = 0.2,
+    min_step_length: float = 0.0,
+) -> xr.DataArray:
+    r"""Compute the sinuosity of a path.
+
+    Sinuosity (S) quantifies the tortuosity of a path by combining
+    turning angle statistics with step-length variability. Higher
+    values indicate more tortuous movement. A perfectly straight
+    path has S = 0.
+    ``
```
Comment (niksirbi) [inline movement/kinematics/path.py]: A stray leftover, which will break docstring rendering.
We can just remove it.

```suggestion
```

File: movement/kinematics/path.py [hunk 17]
```diff
@@ -510,18 +512,181 @@ def compute_path_deviation(
...
+    Returns
+    -------
+    xarray.DataArray
+        An xarray DataArray containing the computed sinuosity,
+        with dimensions matching those of the input data,
+        except ``time`` and ``space`` are removed.
+
+    See Also
+    --------
+    compute_path_length : Total path length between two time points.
```
Comment (niksirbi) [inline movement/kinematics/path.py]: This matches how `compute_path_length` is referenced in othe 'See Also' section in this file.

```suggestion
    compute_path_length : Total distance travelled along a path.
```

File: movement/kinematics/path.py [hunk 18]
```diff
@@ -12,7 +12,9 @@
 import numpy as np
 import xarray as xr
 
-from movement.kinematics.kinematics import compute_backward_displacement
+from movement.kinematics.kinematics import (
+    compute_backward_displacement,
+)
```
Comment (niksirbi) [inline movement/kinematics/path.py]: This import now fits in one line. Could be also fixed by running `pre-commit run --all-files`.
```suggestion
from movement.kinematics.kinematics import compute_backward_displacement
```

File: tests/test_unit/test_kinematics/test_path.py [hunk 19]
```diff
@@ -801,3 +802,112 @@ def test_path_deviation_partially_degenerate_warns(straight_paths):
...
+        assert result.isnull().all()
+    else:
+        xr.testing.assert_allclose(
+            result,
+            xr.full_like(result, expected_value),
+            atol=1e-7,
+        )
+
+
+def test_path_sinuosity_perfect_reversals():
```
Comment (niksirbi) [inline tests/test_unit/test_kinematics/test_path.py]: I would advice deleting this test completely. We also do not need a c̄=−1 guard. It's good to think about edge cases (and I know I've advised doing that), but this may be overthinking it.

This particular edge case requires every step to be identical length to floating-point precision and every turn to be exactly 180°. Real motion tracking data never does this. If that ever happens (extremely unlikely), `inf` isn't even wrong here (we should let the sinuosity index diverge at such a degenerate case).

I would keep the more realistic `test_path_sinuosity_variable_reversals` though (the one that's already passing).

File: tests/test_unit/test_kinematics/test_path.py [hunk 20]
```diff
@@ -801,3 +802,112 @@ def test_path_deviation_partially_degenerate_warns(straight_paths):
...
+# Path sinuosity tests
+# ─────────────────────────────────────────────
+
+
+def test_path_sinuosity_too_few_timepoints(valid_poses_dataset):
+    """Test that sinuosity correctly enforces the minimum length requirement."""
+    # Slice the data to exactly 2 time points (which is only 1 segment)
+    position = valid_poses_dataset.position.isel(time=slice(0, 2))
+
+    with pytest.raises(ValueError):
```
Comment (niksirbi) [inline tests/test_unit/test_kinematics/test_path.py]: Optional: this way we make the test slightly stricter, making sure that the ValueError contains the expected error message.

```suggestion
    with pytest.raises(ValueError, match="3 time points are required"):
```

File: tests/test_unit/test_kinematics/test_path.py [hunk 21]
```diff
@@ -801,3 +802,112 @@ def test_path_deviation_partially_degenerate_warns(straight_paths):
...
+
+
+@pytest.mark.parametrize(
+    "fixture_name, expected_value",
+    [
+        pytest.param("straight_paths", 0.0, id="straight-line"),
+        pytest.param("stationary_paths", np.nan, id="stationary"),
+    ],
+)
+def test_path_sinuosity_known_values(request, fixture_name, expected_value):
```
Comment (niksirbi) [inline tests/test_unit/test_kinematics/test_path.py:878]: I would add at least one more known test case to this test.

You are currently testing straight and stationary paths. It would be good to also verify that a valid non-straight path returns the expected value.

Ideally, we'd need a path fixture with a known closed form solution.

I think somehting like the following should work:

```python
@pytest.fixture
def regular_zigzag_path():
    """Unit-step 90-degree zigzag (Eeast, North, Eeast, North, ...).

    Every step has length 1 (so b = 0) and every turn is 90 degrees
    (so c_bar = cos(90) = 0). Benhamou 2004 Eq. 8 then reduces to
    S = 2 * [p * (1 + c_bar) / (1 - c_bar)]^(-1/2) = 2 / sqrt(1) = 2.0.
    """
    n = 20
    xy = np.zeros((n, 2))
    for i in range(1, n):
        xy[i] = xy[i - 1] + ([1, 0] if i % 2 else [0, 1])
    return xr.DataArray(
        xy,
        dims=["time", "space"],
        coords={"time": np.arange(n), "space": ["x", "y"]},
    )
```

Then you can also include the new fixture in your parametrisation:

```python
@pytest.mark.parametrize(
    "fixture_name, expected_value",
    [
        pytest.param("straight_paths", 0.0, id="straight-line"),
        pytest.param("stationary_paths", np.nan, id="stationary"),
        pytest.param("regular_zigzag_path", 2.0, id="regular-zigzag"),
    ],
)
```

Give it a try and see if that passes.

File: movement/kinematics/path.py [hunk 22]
```diff
@@ -510,18 +512,181 @@ def compute_path_deviation(
...
+    theta = compute_turning_angle(data, min_step_length=min_step_length)
+
+    # --- Summary statistics (NaN-aware) -------------------------------------
+    p_bar = step_lengths.mean(dim="time", skipna=True)
+    c_bar = xr.apply_ufunc(np.cos, theta).mean(dim="time", skipna=True)
+
+    # Guard p_bar == 0 (entirely stationary track) before dividing to prevent
+    # RuntimeWarning. xarray-safe comparison preserves dimension labels.
+    is_stationary = abs(p_bar) < 1e-8
+    p_bar_safe = xr.where(is_stationary, np.nan, p_bar)
```
Comment (niksirbi) [inline movement/kinematics/path.py]: I'm quite condident that we don't neeed this `is_stationary` guard, and also no need to compute `p_bar_safe`. You can delete this entire block and directly use `p_bar` below.

The reason is that, unlike NumPy, xarray doesn't emit 0/0 runtime warnings, and it still gives the correct stationary result (NaN) anyway. 
You can verify this yourself by removing the guard and `p_bar_safe` and rerunning the tests.

File: movement/kinematics/path.py [hunk 23]
```diff
@@ -510,18 +512,181 @@ def compute_path_deviation(
...
+    nan_warn_threshold : float, optional
+        If any point track in the data has at least (:math:`\ge`)
+        this proportion of values missing, a warning will be emitted.
+        Defaults to ``0.2`` (20%).
+    min_step_length : float, optional
+        Minimum step length threshold. Steps shorter than or equal to
+        this value are treated as stationary and excluded from all
+        summary statistics (:math:`\bar{p}`, :math:`\bar{c}`, :math:`b`).
+        Applied symmetrically here and passed to
+        :func:`compute_turning_angle`. Defaults to ``0.0``.
```
Comment (niksirbi) [inline movement/kinematics/path.py]: We've had several discussions about `min_step_length` recently (see issue https://github.com/neuroinformatics-unit/movement/issues/957 and the Community Call [meetings notes from 2026-06-12](https://neuroinformatics.zulipchat.com/#narrow/channel/406001-Movement/topic/Meeting.20Minutes/near/602618163)). Our take on it has changed since the last time I reviewed this PR.

We've decided to remove it from `compute_turning_angle` altogether. Therefore, I no longer think we shoudl expose it here, as it would have to be eventually removed anyway.

I propose removing the `min_step_length` argument, and all sections of the docstring that refer to it.

**Review by niksirbi — CHANGES_REQUESTED** [review]
Hey @isha822, thanks for resolving my earlier comments.

I finally got the chance to have another look at this PR.

To answer your question: I think you should get rid of that 'animal perfectly pacing back-and-forth` test, as I don't see any realistic way for that to happen in real data (see more details in the inline comment).

Apart from that, I'd also advise to drop `min_step_length` from this function completely. Our thinking about it has shifted over time, and we're now gradually moving away from this argument (we also plan to drop it from `compute_turning_angle`). That simplifies things somewhat, see inline comments for details.

After implementing my suggestion, you should run `pre-commit run -a` to fix any lingering formatting problems, and also rebase this branch to the current main. At that point (assuming CI passes), you may mark this PR as ready for review.

I expect we are only 1 review round away from approving this, but keep in mind that I'll be away at a conference all of next week so it may take a while to get back to you.

I appreciate your patience and perseverance.

**Comment by niksirbi** [comment]
Moreover, it looks like #1032 may be merged ahead of this PR. 

If so, we should revise this PR to ensure its consistent with the implementation in #1032 (as the 2 metrics share several commonalities).

**Comment by isha822** [comment]
Thanks so much for the detailed review @niksirbi! ✨

I've just pushed an update that addresses all of the feedback:
- Removed `min_step_length` completely (from the logic, docstrings, and tests).
- Updated the edge-case docstrings to accurately reflect the math handling.
- Added a clipping safeguard to prevent floating-point precision errors.
- Added the regular zigzag unit test.

Marking this as ready for review again. Hopefully it's consistent now. (Checked #1032 )

**Review by niksirbi — APPROVED** [review]
Hey @isha822, I've finally managed to come back to this PR. The implementation was already correct, but I decided to add some cosmetic changes on top, to polish some details in the docstring, comments, and tests. 

You can see my additions as separate commits in the PR's git history. I've also updated the PR description to match its current state.

I'm about to merge this now 🎉  Thanks for all your help and patience.
